### DISCUSSION TITLE
Use of custom logger for controlling test output

### DISCUSSION THREAD

**Discussion by LauSam09** [TanStack/query#5149] [Ideas] [answered]
I recently raised a discussion (#5148) on how to prevent console errors during test runs that involve testing `isError = true` scenarios. 

There seem to be two answers to that:
1. Spy on `console.error` for the relevant tests to suppress console errors. To ensure that only the expected errors are suppressed, we can assert on the spy.
2. Use the deprecated [custom logger](https://tanstack.com/query/v4/docs/react/guides/custom-logger). Although I cannot reproduce this now, I am convinced that at one stage this was generating console warnings itself as this is what initially alerted me to the fact that it is deprecated.

Whilst option 1. is okay, it's going to be relatively frequent noise in tests and feels like a work-around as opposed to a long-term solution.

From the conversation in (#622), and specifically [this comment](https://github.com/TanStack/query/discussions/622#discussioncomment-4565620), it seems as though the custom logger is marked as deprecated because there are no valid use-cases for it, but I think this is one and would like the decision to remove it to be reconsidered.

For reference, here is a minimal reproduction that I added to my original discussion that now makes use of the custom logger.
https://codesandbox.io/p/sandbox/awesome-torvalds-pj2dvj

**Comment by TkDodo** [comment, marked as answer]
the `custom logger` is deprecated because we have removed it in v5. The stance is:
- The custom logger doesn't do anything in production, because it's turned off completely. You also can't enable it.
- The logger was meant to log things developers should pay attention to, not request errors.

so starting with v5, we:

- always log directly with `console.warn` or `console.error`, but only in DEV mode.
- we only log things that you should take attention to, like returning `undefined` from your `queryFn`, which is invalid.
- failing queries will not longer be logged

you can still use the custom logger in v4, but be prepared to just remove it in v5 because you won't be needing it.

**Reply by LauSam09** [reply]
That all sounds good - thank you for the clarification.
